design/src/rfcs/rfc0003_presigning_api.md

```diff
@@ -0,0 +1,192 @@
+  but not all operations can be pre-signed.
+
+**Note:** Pre-signing *needs* to be `async` because the underlying credentials provider used to sign the
+request *may* need to make service calls to acquire the credentials.
+
+Input Pre-signed URL API
+------------------------
+
+Even though generating a pre-signed URL through the fluent client doesn't necessitate an HTTP client,
+it will be clearer that this is the case by allowing the creation of pre-signed URLs directly from an input.
+This would look as follows:
+
+```rust
+let config = aws_config::load_config_from_environment().await;
+let presigning_config = PresigningConfig::expires(Duration::from_secs(86400));
+let presigned_url: String = GetObjectInput::builder()
+    .bucket("example-bucket")
+    .key("example-bucket")
+    .presigned(&config, &presigning_config)
+    .await?;
+```
+
+Creating the URL through the input will exercise the same code path as creating it through the client,
+but it will be more apparent that the overhead of a client isn't present.
+
+Behind the scenes
+-----------------
+
+From an SDK's perspective, the following are required to make a pre-signed URL:
+- Valid request input
+- Endpoint
+- Credentials to sign with
+- Signing implementation
+
+The AWS middleware provides everything except the request, and the request is provided as part
+of the fluent builder API. The generated code needs to be able to run the middleware to fully populate
+a request property bag, but not actually dispatch it. Additionally, the signing headers added by the signing
+implementation will need to be moved into query parameters.
+
+Today, request dispatch looks as follows:
+1. The customer creates a new fluent builder by calling `client.operation_name()`, fills in inputs, and then calls `send()`.
+2. `send()`:
+   1. Builds the final input struct, and then calls its `make_operation()` method with the stored config to create a Smithy `Operation`.
+   2. Calls the underlying Smithy client with the operation.
+3. The Smithy client constructs a Tower Service with AWS middleware and a dispatcher at the bottom, and then executes it.
+4. The middleware acquire and add required signing parameters (region, credentials, endpoint, etc) to the request property bag.
+5. The SigV4 signing middleware signs the request by adding HTTP headers to it.
+6. The dispatcher makes the actual HTTP request and returns the response all the way back up the Tower.
+
+Pre-signing will take advantage of a lot of these same steps, but will cut out the `Operation` and
+replace the dispatcher with a pre-signed URL generator:
+1. The customer creates a new fluent builder by calling `client.operation_name()`, fills in inputs, and then calls `presigned()`.
+2. `presigned()`:
+   1. Builds the final input struct, and then calls a new `make_request()` method with the stored config.
+   2. Constructs a Tower Service with `AwsMiddleware` layered in, and a `PresignedUrlGeneratorLayer` at the bottom.
```

Comment:
the signing process is actually totally different—you can't take the headers and move them to the URI, you'll end up with the wrong signature. This doesn't invalidate this design, but we'll need to plan to either drop in a different SigV4 stage or drive this with signing configuration.

(And allocate time to add support to the signer)

Reply:
It is an interesting strategy to use the tower stack to generate the pre-signed URLs. Tower does come with non-trivial complexity, vs. extracting necessary methods and calling them directly. Does going through tower here provide enough benefit to use it? Do we expect customers to inject custom middleware when generating presigned URLs?

Reply:
From my perspective, it is easier to replace the bottom of the Tower stack than to create a manual call chain to accomplish the same thing. Thinking about it more though, it seems like this will also make it easier to write service customizations in the future since we just need to add a customization middleware layer, and then it should work for both requests and pre-signed URLs. Regardless, we should be able to change course if it doesn't work out that way since this isn't part of the public API.